doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,291 @@
+- Documentation is required to demonstrate how to use indicators for each approved cryptographic algorithm.
+- Testing is required to execute all services and verify that the indicator provides an unambiguous indication of whether the service utilizes an approved cryptographic algorithm, security function or process in an approved manner or not.
+- The Security Policy may require updates related to indicators.
+AWS/google have added a table in their security policy called
+‘Non-Approved Algorithms not allowed in the approved mode of operation’.
+An example is RSA with a keysize of < 2048 bits (which has been enforced by [4]).
+
+### Legacy Support
+
+Due to key transitions [4] we may have some legacy algorithms that are in a state of only being approved for processing (verification, decryption, validation), and not for protection (signing, encrypting, keygen).
+For example DSA.
+
+The options are:
+- Completely remove the algorithm from the FIPS provider.
+  This is simple but means older applications can no longer process existing data, which is not ideal.
+- Allow the algorithm but make it not approved with an context specific indicator.
+It is safer to make the protection operations fail rather than use an indicator.
+The processing operation for DSA would set the indicator to approved.
+
+
+### Security Checks
+
+OpenSSL currently defines configurable FIPS options.
+These options are supplied via the FIPS configuration file - which is normally setup via fipsinstall.
+
+- FIPS_FEATURE_CHECK(FIPS_security_check_enabled, fips_security_checks, 1)
+- FIPS_FEATURE_CHECK(FIPS_tls_prf_ems_check, fips_tls1_prf_ems_check, 0)
+- FIPS_FEATURE_CHECK(FIPS_restricted_drbg_digests_enabled, 0)
+- OSSL_PROV_FIPS_PARAM_CONDITIONAL_ERRORS selftest_params.conditional_error_check
+
+The following functions are available in providers/common/security_check.c.
+
+- ossl_rsa_check_key()
+- ossl_ec_check_key()
+- ossl_dsa_check_key()
+- ossl_dh_check_key()
+- ossl_digest_get_approved_nid_with_sha1()
+- ossl_digest_is_allowed()
+
+Anywhere where these functions are called an indicator MAY be required.
+Because these options are available I do not think it is sufficient to
+document this in the security policy.
+
+Each of these functions contains code of the following form:
+~~~
+#if !defined(OPENSSL_NO_FIPS_SECURITYCHECKS)
+    if (ossl_securitycheck_enabled(ctx)) {
+      // Do some checks, and maybe return 0 for a hard failure
+      ...
+    }
+}
+~~~
+
+OPENSSL_NO_FIPS_SECURITYCHECKS is also a configurable option
+If the security checks are not enabled then it is unapproved?
+
+
+## Implementation options
+
+The above requirements indicate 2 options.
+
+### Option 1
+- Dont allow ANY non approved algorithms and then a indicator is not required.
```

Comment:
Blank line above :)

This is the approach of KeyPair's 140-3 validation.  It is not ideal.